### PR TITLE
Adds button to reset all ToDos on the page

### PR DESCRIPTION
**Pull request by RobertWeinmeister**
This change allows you to reset all checkboxes on a page with the press of a button. The reset is stored as a single edit in the page history. This is very helpful if you need to uncheck all ToDos regularly as you don't need to click them all separately and it doesn't clutter the edit history.

This covers also the uncheck bit of #158.

## Usage

Add `~~TODO:UNCHECKALL~~` to the page. This generates a button with the label "Uncheck all todos".
image
Once clicked, all ToDos of this page are reset.
image
A minor edit with the summary "Unchecked all ToDos" is created.

### syntax.php

Adds the new pattern as
`const TODO_UNCHECK_ALL = '~~TODO:UNCHECKALL~~';`
and replaces it with the button for the generated page:
`$renderer->doc .= '<button type="button" class="todouncheckall">Uncheck all todos</button>';`

### script.js

The uncheck function is bound to the click event of this button:
```
jQuery('button.todouncheckall').click(function () {
        ToDoPlugin.uncheckall();
    });
```
The function then leads to an ajax call:
```
        jQuery.post(
            DOKU_BASE + 'lib/exe/ajax.php',
            {
                call: 'plugin_todo',
                mode: 'uncheckall',
                pageid: jQuery('input.todocheckbox:first').data().pageid
            },
        whenCompleted, 'json');
```

To distinguish between the click of a ToDo and this button, a new parameter `mode` is added:
`mode: 'checkbox'` & `mode: 'uncheckall'`

On completion, the current page is also modified to reset all ToDos:
```
var whenCompleted = function () {
            jQuery('input.todocheckbox').each(function() {
                jQuery(this).prop('checked', false);
            });

            jQuery('span.todoinnertext').each(function () {
                if (jQuery(this).parent().is("del")) {
                    jQuery(this).unwrap();
                }
            });

            jQuery('span.todouser').each(function () {
                jQuery(this).remove();
            });
            
            ToDoPlugin.locked = false;
        };
```
This is only for an immediate feedback. The permanent change is the one from the ajax call.

### action.php

Includes a slight rewrite of the way the parameters are read to adapt it to the introduction of modes.

The actually relevant addition is the replacement of the checked markers with regex, # and any following non-whitespace character is removed:
`$newWikitext = preg_replace('/(<todo.*?)(\s+#[^>\s]*)(.*?>|\s.*?<\/todo>)/', '$1$3', $wikitext);`

The new wiki text is then saved as a minor edit:
`saveWikiText($ID, $newWikitext, 'Unchecked all ToDos', $minoredit = true);`

## Testing

I tested this with a lot of possible combinations of attributes (@, due, start).

## Possible improvements

The label of the new button and the edit summary are hard-coded in English. Translations for all the currently available languages could be added.

### CODE REVIEW

**Comment by AdosJ** [comment]
This needs to be merged. Thank you for your work! :)

**Comment by runout-at** [comment]
Looks like there is a bug...

It seems that the _handler_ function doesn't return all the array keys which are expected after this commit - depending on $state switch.

@RobertWeinmeister how do you prefer to solve this? In which of the functions? Possibly _handler_ should have a consistent return value?
